### PR TITLE
New decoupling layout parameter

### PR DESCRIPTION
**Pull request by BenoitJeanson**
**Please check if the PR fulfills these requirements** *(please use `'[x]'` to check the checkboxes, or submit the PR and then click the checkboxes)*
- [x] The commit message follows our guidelines
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

**Does this PR already have an issue describing the problem ?** *If so, link to this issue using `'#XXX'` and skip the rest*
no

**What kind of change does this PR introduce?** *(Bug fix, feature, docs update, ...)*
decoupling of subpackage from model by decoupling LayoutParameters and LayoutContext to avoid circular dependencies of subpackages.
(if any of the questions/checkboxes don't apply, please delete them entirely)

### CODE REVIEW

File: single-line-diagram-core/src/main/java/com/powsybl/sld/layout/PositionVoltageLevelLayout.java [hunk 1]
```diff
@@ -88,17 +92,31 @@ private void adaptPaddingToSnakeLines(LayoutParameters layoutParam) {
...
 
     private void calculateCellCoord(VoltageLevelGraph graph, LayoutParameters layoutParam) {
-        graph.getBusCellStream().forEach(cell -> cell.calculateCoord(layoutParam, createLayoutContext(graph, cell, layoutParam)));
-        graph.getShuntCellStream().forEach(cell -> cell.calculateCoord(layoutParam, null));
+        CalculateCellCoordVisitor calculateCellCoordVisitor = new CalculateCellCoordVisitor(layoutParam);
+        graph.getBusCellStream().forEach(cell -> {
+            calculateCellCoordVisitor.setLayoutContext(createLayoutContext(graph, cell, layoutParam));
+            cell.accept(calculateCellCoordVisitor);
+        });
+        calculateCellCoordVisitor.setLayoutContext(null);
```
Comment (flo-dup) [inline single-line-diagram-core/src/main/java/com/powsybl/sld/layout/PositionVoltageLevelLayout.java:113]: You'd rather use a new `CalculateCellCoordVisitor` for each bus cell and add a constructor with LayoutContext argument

**Review by flo-dup — CHANGES_REQUESTED** [review]
That's really nice work, grouping the calculation code in dedicated visitors makes it more readable - thanks! A few minor comments before merging it

File: single-line-diagram-core/src/main/java/com/powsybl/sld/layout/BlockCalculationVisitors.java [hunk 2]
```diff
@@ -0,0 +1,244 @@
...
+    public BlockCalculationVisitors(LayoutParameters layoutParameters) {
+        this.layoutParameters = layoutParameters;
+    }
+
+    public CalculateCellHeightVisitor createCalculateCellHeight(Set<Node> encounteredNodes) {
+        return new CalculateCellHeightVisitor(encounteredNodes);
+    }
+
+    public CalculateCellHeightVisitor createCalculateCellHeight() {
+        return new CalculateCellHeightVisitor(new HashSet<>());
```
[the remark this answers is not in the document]
Comment (BenoitJeanson) [reply single-line-diagram-core/src/main/java/com/powsybl/sld/layout/BlockCalculationVisitors.java:38]: ok - empty constructor removed

File: single-line-diagram-core/src/main/java/com/powsybl/sld/layout/BlockCalculationVisitors.java [hunk 3]
```diff
@@ -0,0 +1,244 @@
...
+    }
+
+    public CalculateCoord createCalculateCoord(LayoutContext layoutContext) {
+        return new CalculateCoord(layoutContext);
+    }
+
+    public class CalculateCellHeightVisitor implements BlockVisitor {
```
[the remark this answers is not in the document]
Comment (BenoitJeanson) [reply single-line-diagram-core/src/main/java/com/powsybl/sld/layout/BlockCalculationVisitors.java:45]: ok

File: single-line-diagram-core/src/main/java/com/powsybl/sld/layout/BlockCalculationVisitors.java [hunk 4]
```diff
@@ -0,0 +1,244 @@
...
+            // already encountered
+            long nbNodes = block.getNodes().stream().filter(n -> !encounteredNodes.contains(n) && n.getType() != BUS)
+                    .count();
+
+            this.blockHeight = (nbNodes - 1) * componentHeight;
+        }
+
+        @Override
+        public void visit(SerialBlock block) {
+            calculateSubHeight(block, (a, b) -> a + b);
```
[the remark this answers is not in the document]
Comment (BenoitJeanson) [reply single-line-diagram-core/src/main/java/com/powsybl/sld/layout/BlockCalculationVisitors.java:89]: ok

File: single-line-diagram-core/src/main/java/com/powsybl/sld/layout/BlockCalculationVisitors.java [hunk 5]
```diff
@@ -0,0 +1,244 @@
...
+        public void visit(BodyParallelBlock block) {
+            calculateSubHeight(block, Math::max);
+        }
+
+        @Override
+        public void visit(UndefinedBlock block) {
+            calculateSubHeight(block, Math::max);
+        }
+
+        private void calculateSubHeight(ComposedBlock block, DoubleBinaryOperator op) {
```
[the remark this answers is not in the document]
Comment (BenoitJeanson) [reply single-line-diagram-core/src/main/java/com/powsybl/sld/layout/BlockCalculationVisitors.java:107]: ok

File: single-line-diagram-core/src/main/java/com/powsybl/sld/layout/BlockCalculationVisitors.java [hunk 6]
```diff
@@ -0,0 +1,244 @@
...
+            blockHeight = 0.;
+            for (int i = 0; i < block.getSubBlocks().size(); i++) {
```
[the remark this answers is not in the document]
Comment (BenoitJeanson) [reply single-line-diagram-core/src/main/java/com/powsybl/sld/layout/BlockCalculationVisitors.java:109]: ok

File: single-line-diagram-core/src/main/java/com/powsybl/sld/layout/CalculateCellCoordVisitor.java [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+
+import static com.powsybl.sld.model.coordinate.Position.Dimension.*;
+import static com.powsybl.sld.model.coordinate.Coord.Dimension.*;
+import static com.powsybl.sld.model.blocks.Block.Extremity.*;
+
+/**
+ * @author Benoit Jeanson <benoit.jeanson at rte-france.com>
+ */
+public class CalculateCellCoordVisitor implements CellVisitor {
+    private LayoutParameters layoutParameters;
```
[the remark this answers is not in the document]
Comment (BenoitJeanson) [reply single-line-diagram-core/src/main/java/com/powsybl/sld/layout/CalculateCellCoordVisitor.java:27]: ok

File: single-line-diagram-core/src/main/java/com/powsybl/sld/layout/CalculateCellCoordVisitor.java [hunk 8]
```diff
@@ -0,0 +1,131 @@
...
+    }
+
+    @Override
+    public void visit(ExternCell cell) {
+        calculateRootCoord(cell.getRootBlock(), layoutContext);
+    }
+
+    @Override
+    public void visit(ShuntCell cell) {
+        if (cell.getRootBlock() instanceof BodyPrimaryBlock) {
```
[the remark this answers is not in the document]
Comment (BenoitJeanson) [reply single-line-diagram-core/src/main/java/com/powsybl/sld/layout/CalculateCellCoordVisitor.java:56]: ok

File: single-line-diagram-core/src/main/java/com/powsybl/sld/layout/PositionVoltageLevelLayout.java (hunk 1 above)
Comment (BenoitJeanson) [reply single-line-diagram-core/src/main/java/com/powsybl/sld/layout/PositionVoltageLevelLayout.java:113]: ok

**Review by flo-dup — APPROVED** [review]

**Pull request merged**